components/skin/ui/src/main/resources/PhenoTips/Skin.xml

```diff
@@ -507,39 +507,144 @@ document.observe('xwiki:suggest:containerCreated', function(event) {
+        // differences with the state of the form when navigating away even if the user
+        // hasn't actually changed any data.
+        // Setting a timeout, even zero, delays the initialization until other
+        // form elements have been modified already.
+        setTimeout(recordFormData, 0);
+        // Record the state of the form again once the form is saved.
+        document.observe('xwiki:document:saved', recordFormData);
+        $$("input[name^='action_']").invoke('observe', 'click', recordFormData);
+        // Upon navigating away, check if anything has changed in the form and warn the user
+        // if it did.
+        window.addEventListener('beforeunload', function(event) {
+          if (_this.isFormDirty()) {
+            var message = "$!escapetool.javascript($services.localization.render('phenotips.navigation.beforeunload.warning'))";
+            (event || window.event).returnValue = message;
+            return message;
+          }
+          return undefined;
+        });
+
+        // Set up the smart autosave
+        if (this.config.autosave.enabled) {
+          this.timer = new PeriodicalExecuter(this.saveIfFormDirty.bind(this), this.config.autosave.frequency);
+        }
+      }
+    },
+    isFormDirty : function() {
+      return this.form ? (this.formData != this.form.serialize()) : false;
+    },
+    saveForm : function(successCallback, failureCallback) {
+      document.fire("xwiki:actions:save", {"continue": true, "form" : this.form});
+      // If there's a callback for there's no unsaved data:
+      if (typeof successCallback == 'function') {
+        // Call the callback function
+        document.observe("xwiki:document:saved", function onSaved() {
+          document.stopObserving("xwiki:document:saved", onSaved);
+          successCallback();
+        });
+        // At the same time, if there's NO callback for when the saving failed, ask user if it's OK to proceed anyway with the successCallback
+        if (typeof failureCallback != 'function') {
+          document.observe("xwiki:document:saveFailed", function onSaveFailed() {
+            document.stopObserving("xwiki:document:saveFailed", onSaveFailed);
+            (new XWiki.widgets.ConfirmationBox({
+              onYes: successCallback,
+              onNo: function(){}
+            },
+            {
+              confirmationText: "$!escapetool.javascript($services.localization.render('phenotips.save.failed.confirmProceed'))", // "Failed to save the latest modifications. Proceed anyway?"
+              showCancelButton: false
+             })).showDialog();
+          });
         }
-        return undefined;
+      }
+      // However, if there is a callback for when the saving failed, call that one at the saveFailed event
+      (typeof failureCallback == 'function') &amp;&amp; document.observe("xwiki:document:saveFailed", function onSaveFailed() {
```

Comment:
If the save succeeds this time, then `failureCallback` listeners accumulate, and a subsequent save failure could call an unrelated `failureCallback` registered earlier. Same for a temporary failure which fails to unregister a `successCallback`. Both on `xwiki:document:saved` and `xwiki:document:saveFailed` all event listeners should be cleared.